kquery/KQuery.g4

```diff
@@ -0,0 +1,357 @@
+     All rights reserved.
+     Redistribution and use in source and binary forms, with or without
+     modification, are permitted provided that the following conditions
+     are met:
+     1. Redistributions of source code must retain the above copyright
+        notice, this list of conditions and the following disclaimer.
+     2. Redistributions in binary form must reproduce the above copyright
+        notice, this list of conditions and the following disclaimer in the
+        documentation and/or other materials provided with the distribution.
+     3. The name of the author may not be used to endorse or promote products
+        derived from this software without specific prior written permission.
+     THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+     IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+     OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+     IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+     INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+     NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+     DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+     THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+     (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+     THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+// Grammar for KLEE KQuery parsing. (A bit more verbose with richer parsing)
+// Ported to Antlr4 by Sumit Lahiri.
+grammar KQuery;
+
+kqueryExpression 
+    : queryStatements EOF
+    ;
+
+queryStatements
+    : (ktranslationUnit)*
+    ;
+    
+ktranslationUnit
+    : arrayDeclaration
+    | queryCommand
+    ;
+
+queryCommand 
+    : LeftParen Query evalExprList queryExpr RightParen 
+    ;
+        
+queryExpr 
+    : expression                                #SingletonQueryExpr
+    | expression evalExprList                   #WithEvalExpr
+    | expression evalExprList evalArrayList     #WithEvalExprAndArrayList
+    ;
+    
+evalExprList 
+    : LeftBracket expressionList RightBracket  
+    ;
+
+evalArrayList 
+    : LeftBracket identifierList RightBracket  
+    ;
+
+expressionList : (expression)*;
+identifierList : (Identifier)*;
+
+arrayDeclaration
+    : Array arrName LeftBracket numArrayElements RightBracket 
+        Colon domain Arrow rangeLimit Equal arrayInitializer
+    ;
+    
+numArrayElements
+    : Constant
+    ;
+    
+arrayInitializer
+    : Symbolic 
+    | LeftBracket numberList RightBracket
+    ;
+    
+expression
+    : varName                                                                               #VariableName
+    | namedConstant Colon expression                                                        #NamedAbbreviation
+    | LeftParen widthOrSizeExpr number RightParen                                           #SizeQuery
```

Comment:
This needs to be refactored, e.g., `expression: varName | ... | LeftParen (withOrSizeExpr number | arithmeticExpr widthOrSizeExpr leftExpr rightExpr  | .... RightParen | ....;` Your grammar is quite inefficient.